### ISSUE TITLE
Counterintuitive (wrong?) parenthesis when combining exp() and powers

### ISSUE THREAD

**Issue by jonbarron** [google/latexify_py#189]
## Description

When you latexify something like log((x/c)^2), the rendering you get looks (to me at least) like it's log(x/c)^2:
<img width="333" alt="image" src="https://github.com/google/latexify_py/assets/889526/c4c7a979-d5f1-459c-9957-daecbf6fd405">

In contrast, Wolfram Alpha for example will insert extra parens to disambiguate this case.
<img width="292" alt="image" src="https://github.com/google/latexify_py/assets/889526/28311362-6ed8-4911-a572-48cdee621280">

## Reproduction

Describe how to reproduce the issue by other people.

```
@latexify.function()
def f(x, c):
  return np.log((x/c)**2)
f
```

**Comment by odashi** [comment]
Yeah I know this kind of ambiguity still remains around unary operators, including bracket-less functions as you reported above. We haven't figure out the correct rules around this yet.
Could you provide more examples that we should treat?

**Comment by xiziqiao** [comment]
Hi, there. I am really interested in this project. Could I participate in solving this minor issue?

**Comment by odashi** [comment]
@xiziqiao Yeah please describe the ideas if you have it already.

I want to avoid simply adding brackets to all functions because of its redundancy, but currently I have no clear policy to appropriately choose whether add a bracket or not.

**Comment by xiziqiao** [comment]
@odashi Hi Mr. Odashi, thanks for responding. Our thoughts now is pretty straightforward. If we have a input like log x^2, then the output latex should be (log x)^2. If we want the latex output to be log(x^2), then the python input should be log（（x^2））.
Did I illustrate my idea clearly? We can communicate by email if you have any questions. Additional thoughts and advice are much appreciated. My email would be xiziqiao@umich.edu.

**Comment by odashi** [comment]
@xiziqiao Please keep discussion on this thread to avoid hidden decisions.

The library has several rules to avoid parentheses around the argument of several functions to follow the normal math notation. That means the abbreviation *is* the special case (=most functions are exactly followed by the parentheses), and this issue is eventually related to whether re-considering this rule or not.

<img width="639" alt="無題" src="https://github.com/google/latexify_py/assets/1023695/2a66adef-1ad0-45f0-b5ff-27088208c079">

**Comment by odashi** [comment]
I didn't remember soon about the context around the pow operator, but should it be treated similarly to that of other binary operators?

**Comment by tesknight** [comment]
how about just add a parentheses when it is found a pow expression is taken as arg in a function call? I think only the pow operator need a special treatment as its precedence is higher than unary function call

**Comment by xiziqiao** [comment]
@odashi I am thinking the same thing as @tesknight. We can simply add a parenthese when pow function is put into the function call.

**Comment by odashi** [comment]
It looks the cause comes from the custom precedence of function calls. It is set as a smaller value than pow, resulting in removal of brackets between a unary function and pow.
I think this rule can be removed so that unary functions of non-constant always require brackets.

**Comment by odashi** [comment]
Yeah removing the rule works:

![Uploading 無題.png…]()

**Comment by odashi** [comment]
Yeah removing the rule works:
<img width="407" alt="無題" src="https://github.com/google/latexify_py/assets/1023695/6e216a25-ea5c-4cd4-b0dd-b053cf445c21">

**Comment by odashi** [comment]
Unfortunately it brings another bug: 

<img width="403" alt="無題" src="https://github.com/google/latexify_py/assets/1023695/e78da1aa-df3a-475d-9395-e268789ed103">

**Comment by xiziqiao** [comment]
Okay. This seems neat. Are you going to keep changing this rule? If so I will take a look of this new bug

**Comment by odashi** [comment]
I think the above change shouldn't be applied, and we should change the behavior around the generation rules of Call instead:

https://github.com/google/latexify_py/blob/1cfb8e78527cb8e13dc647e9c8d91f789a427402/src/latexify/codegen/expression_codegen.py#L409-L421

I'm working on this, will propose some appropriate changes.

**Comment by tesknight** [comment]
I agree. Simply make pow a higher precedence or unary function call a higher precedence both bring some bugs. We may need to force wrap it when a Pow expression be its arg

**Comment by tesknight** [comment]
I think we can make pow and call have the same precedence and let force_wrap work when they were taken as arg of each other.
I make such changes and it pass the examples you provide above:
<img width="1057" alt="截屏2023-11-18 下午3 35 49" src="https://github.com/google/latexify_py/assets/65383033/72aadeb9-fead-47de-b923-b9756e2e4380">

**Comment by odashi** [comment]
I made a similar change: https://github.com/google/latexify_py/pull/195
The difference is that I just added the pow flag in `visit_Call`, and removed the same-precedence rule in force_wrap since I found that this is not actually effective in the current implementation.
